### ISSUE TITLE
Problem with using subsample.particle_properties

### ISSUE THREAD

**Issue by hooteoos-waltz** [hainegroup/oceanspy#62]
* OceanSpy version:

### Description

I tried to extract properties at the location of particles using ospy.subsample.particle_properties. But then I got an error:
ValueError: cannot reindex or align along dimension 'time' because the index has duplicate values
Also, since the code has changed, how/where should I specify which variable from the dataset do I want to extract at the location of particles? The od that is passed to the function should already be a cutted dataset that include only my variable of interest?

### What I Did

```
import oceanspy as ospy
import xarray as xr
import numpy as np
import scipy.io as sio
od_rect = ospy.open_oceandataset.EGshelfIIseas2km_ASR()
particle_data=sio.loadmat('particlefile.mat')
# saving variables from the matrix
All_lons = particle_data['f_lons'].transpose()
All_lats = particle_data['f_lats'].transpose()
All_deps = particle_data['f_deps'].transpose()
timee=particle_data['timee'].transpose()

ds_lagr = ospy.subsample.particle_properties(od_rect, 
                                        times          = squeezed_timee,
                                        Ypart          = All_lons,
                                        Xpart          = All_lats,
                                        Zpart          = All_deps)

The errors:
Cutting out the oceandataset.
Extracting Eulerian properties of particles.
/home/idies/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: UserWarning: 
Particle properties extraction drops variables on `time_midp` dimension. 
Dropped variables: ['time_midp'].
  """

ValueErrorTraceback (most recent call last)
<ipython-input-51-2f70f56633f6> in <module>
      3                                         Ypart          = All_lons,
      4                                         Xpart          = All_lats,
----> 5                                         Zpart          = All_deps)

~/miniconda3/lib/python3.6/site-packages/oceanspy/subsample.py in particle_properties(od, times, Ypart, Xpart, Zpart, **kwargs)
   1025                              dims   = {dim})
   1026         itmp = itmp.sel({dim: tmp}, method='nearest')
-> 1027         i_ds['i'+dim] = itmp
   1028 
   1029     # Convert 2 cartesian

~/miniconda3/lib/python3.6/site-packages/xarray/core/dataset.py in __setitem__(self, key, value)
    987                                       'to set Dataset values')
    988 
--> 989         self.update({key: value})
    990 
    991     def __delitem__(self, key):

~/miniconda3/lib/python3.6/site-packages/xarray/core/dataset.py in update(self, other, inplace)
   2528         """
   2529         inplace = _check_inplace(inplace, default=True)
-> 2530         variables, coord_names, dims = dataset_update_method(self, other)
   2531 
   2532         return self._replace_vars_and_dims(variables, coord_names, dims,

~/miniconda3/lib/python3.6/site-packages/xarray/core/merge.py in dataset_update_method(dataset, other)
    578 
    579     return merge_core([dataset, other], priority_arg=1,
--> 580                       indexes=dataset.indexes)

~/miniconda3/lib/python3.6/site-packages/xarray/core/merge.py in merge_core(objs, compat, join, priority_arg, explicit_coords, indexes)
    433 
    434     coerced = coerce_pandas_values(objs)
--> 435     aligned = deep_align(coerced, join=join, copy=False, indexes=indexes)
    436     expanded = expand_variable_dicts(aligned)
    437 

~/miniconda3/lib/python3.6/site-packages/xarray/core/alignment.py in deep_align(objects, join, copy, indexes, exclude, raise_on_invalid)
    215 
    216     aligned = align(*targets, join=join, copy=copy, indexes=indexes,
--> 217                     exclude=exclude)
    218 
    219     for position, key, aligned_obj in zip(positions, keys, aligned):

~/miniconda3/lib/python3.6/site-packages/xarray/core/alignment.py in align(*objects, **kwargs)
    163             new_obj = obj.copy(deep=copy)
    164         else:
--> 165             new_obj = obj.reindex(copy=copy, **valid_indexers)
    166         new_obj.encoding = obj.encoding
    167         result.append(new_obj)

~/miniconda3/lib/python3.6/site-packages/xarray/core/dataset.py in reindex(self, indexers, method, tolerance, copy, **indexers_kwargs)
   1911         variables = alignment.reindex_variables(
   1912             self.variables, self.sizes, self.indexes, indexers, method,
-> 1913             tolerance, copy=copy)
   1914         coord_names = set(self._coord_names)
   1915         coord_names.update(indexers)

~/miniconda3/lib/python3.6/site-packages/xarray/core/alignment.py in reindex_variables(variables, sizes, indexes, indexers, method, tolerance, copy)
    321                 raise ValueError(
    322                     'cannot reindex or align along dimension %r because the '
--> 323                     'index has duplicate values' % name)
    324 
    325             target = utils.safe_cast_to_index(indexers[name])

ValueError: cannot reindex or align along dimension 'time' because the index has duplicate values

```

**Comment by malmans2** [comment]
Is this fixed or should I look into it?

**Comment by hooteoos-waltz** [comment]
Yes! Check out the updated script and the error in the shared folder. I copied it here as well.
```
import oceanspy as ospy
import xarray as xr
import matplotlib.pyplot as plt
import numpy as np
import scipy.io as sio
import datetime
od_rect = ospy.open_oceandataset.EGshelfIIseas2km_ASR()
particle_data=sio.loadmat('Feb29_particles.mat')
All_lons = particle_data['f_lons'].transpose()
All_lats = particle_data['f_lats'].transpose()
All_deps = particle_data['f_deps'].transpose()
#timee=particle_data['timee'].transpose()
p_times=np.arange('2008-02-29T00:00', '2008-03-30T06:00', dtype='datetime64[6h]')
ds_lagr = ospy.subsample.particle_properties(od_rect, 
                                        times          = p_times,
                                        Ypart          = All_lons[:,10:20],
                                        Xpart          = All_lats[:,10:20],
                                        Zpart          = All_deps[:,10:20],
                                        varList='Temp')
fig, ax = plt.subplots()
ax = ds_lagr.plot.time_series(varName='Temp')
```

**Comment by malmans2** [comment]
Using all particles took 8s:
```
All_lons = particle_data['f_lons'].transpose()
All_lats = particle_data['f_lats'].transpose()
All_deps = particle_data['f_deps'].transpose()
p_times=np.arange('2008-02-29T00:00', '2008-03-30T06:00', dtype='datetime64[6h]')
ds_lagr = ospy.subsample.particle_properties(od_rect, 
                                             times          = p_times,
                                             Ypart          = All_lats,
                                             Xpart          = All_lons,
                                             Zpart          = All_deps,
                                             varList='Temp')
fig, ax = plt.subplots()
ax = ds_lagr.plot.time_series(varName='Temp', add_legend=False)
```
![particles](https://user-images.githubusercontent.com/22245117/54470298-de70e180-477b-11e9-94d9-4ca508a36a50.png)

**Comment by hooteoos-waltz** [comment]
Oops! Yeah I switch the YPart and XPart and I do not get any error. But for some reason it does not do the plotting. It only says: "Cutting out the oceandataset." without error and does not plot anything. Is that a plot issue? just FYI, I am using Jupiter notebook.
```
# Plot
fig, ax = plt.subplots()
ax = ds_lagr.plot.time_series(varName='Temp', add_legend=False)
```

**Comment by hooteoos-waltz** [comment]
I just realized I can't do any plotting! when I run the test plot (below) it give : [<matplotlib.lines.Line2D at 0x7fa491a19748>] as an output and does not show the plot. Is that a sciserver issue? does anyone else have this problem?

```
# A test plot
x = np.arange(0.0, 2.0, 0.01)
y = 1 + np.sin(2 * np.pi * t)

fig, ax = plt.subplots()
ax.plot(x, y)
```

**Comment by hooteoos-waltz** [comment]
And a quick edit: y should be:  ``` y = 1 + np.sin(2 * np.pi * x)``` and not t

**Comment by malmans2** [comment]
For testing reasons matplotlib inline is not active by default. 
Just add %matplotlib inline in your first cell.

**Comment by malmans2** [comment]
If you are using my demo: https://oceanspy.readthedocs.io/en/latest/Demo2_Subsample.html
Also, a quick perfomance tip. In my demo I'm requesting Sigma0 directly in the subsampling dunction for particle properties. This would compute Sigma0 for the whole domain before actually extracting T and S. If you are using a large domain (demo was just surface), you should extract T and S first, then compute Sigma0 using OceanSpy. If you use the syntax that adds Sigma0 to the oceandataset (`od_part.compute.potential_density_anomaly()`), then you can also use OceanSpy plots and animations.

**Comment by hooteoos-waltz** [comment]
Great! Adding "%matplotlib inline" worked!
 The only issue is that I tried different variables; 'V' worked, but 'momVort3' did not. The subsampling didn't give any error, but I tried plotting it by "ax = ds_lagr.plot.time_series(varName='momVort3', add_legend=False)" and it crashed the Kernel. I tried it a couple of times and the same thing happened. The kernel restarted.


Thanks for the comment on the potential density. I agree it's better/faster to compute T&S along the track and then compute the density. Although I think the results wouldn't be exactly the same as computing the density field and the use the get properties function (because equation of state is non-linear).

**Comment by malmans2** [comment]
1. Not sure way, but it looks like a performance issue. I'll look into it.
2. The result would be exactly the same. I'm always using the same (non-linear, but it doesn't really matter) equation, so for any given T-S the resulting Sigma0 will always be the same.

**Comment by hooteoos-waltz** [comment]
1.  The same thing happens when it wants to: print(ds_lagr.dataset.momVort3.values) . Maybe somewhere it becomes infinity or undetermined?

2. Although in both cases the same equation of states is used, the input T&S to the equation of state is different: The first one is the T&S at the grid points, the second one is the T&S that is already interpolated to the position of particles. So the answer is slightly different. But it might not be important. (and we already use the  second method for the particles)

**Comment by malmans2** [comment]
Both the Lagrangian particle code that you are using and OceanSpy use nearest-neighbor interpolation (https://oceanspy.readthedocs.io/en/latest/generated/oceanspy.subsample.particle_properties.html). So for both OceanSpy and the particle code it doesn't matter if you compute the potential density anomaly on the whole field, or just at the specific location of the particles.

**Comment by malmans2** [comment]
OK, I found the bottleneck. I'll make a pull request later today.

**Comment by malmans2** [comment]
The issue should be fixed in this branch: https://github.com/malmans2/oceanspy/tree/complete_tests .
Please let me know if you're able to extract any field (U, V, G, C grid).

**Comment by hooteoos-waltz** [comment]
Is this on the main branch? I updated the ospy by: pip install git+https://github.com/malmans2/oceanspy.git and tried again and the same problem with extracting vorticity happened. Should I install that branch in a different way?

**Comment by hooteoos-waltz** [comment]
I used  git clone --single-branch --branch complete_tests https://github.com/malmans2/oceanspy.git && pip install -e oceanspy/ and then tried again to calculate the 3rd component of vorticity and the same thing happened. Maybe the branch is not properly installed?

**Comment by malmans2** [comment]
Should be fixed here: #63
